paz/models/classification/cnn2Plus1.py

```diff
@@ -0,0 +1,354 @@
+"""
+
+
+class Conv2Plus1D(Layer):
+    def __init__(self, filters, kernel_size, padding):
+        """
+        A sequence of convolutional layers that first apply the convolution operation over the
+        spatial dimensions, and then the temporal dimension.
+        """
+        super().__init__()
+        initializer_glorot_spatial = tf.keras.initializers.GlorotUniform(seed=random.randint(0, 1000000))
+        initializer_glorot_temporal = tf.keras.initializers.GlorotUniform(seed=random.randint(0, 1000000))
+
+        self.seq = Sequential([
+            # Spatial decomposition
+            Conv3D(filters=filters,
+                   kernel_size=(1, kernel_size[1], kernel_size[2]),
+                   padding=padding,
+                   kernel_initializer=initializer_glorot_spatial),
+            # Temporal decomposition
+            Conv3D(filters=filters,
+                   kernel_size=(kernel_size[0], 1, 1),
+                   padding=padding,
+                   kernel_initializer=initializer_glorot_temporal)
+        ])
+
+    def call(self, x):
+        return self.seq(x)
+
+
+class ResidualMain(Layer):
+    """
+    Residual block of the model with convolution, layer normalization, and the
+    activation function, ReLU.
+    """
+
+    def __init__(self, filters, kernel_size):
+        super().__init__()
+        self.seq = Sequential([
+            Conv2Plus1D(filters=filters,
+                        kernel_size=kernel_size,
+                        padding='same'),
+            LayerNormalization(),
+            ReLU(),
+            Conv2Plus1D(filters=filters,
+                        kernel_size=kernel_size,
+                        padding='same'),
+            LayerNormalization()
+        ])
+
+    def call(self, x):
+        return self.seq(x)
+
+
+class Project(Layer):
+    """
+    Project certain dimensions of the tensor as the data is passed through different
+    sized filters and downsampled.
+    """
+
+    def __init__(self, units):
+        super().__init__()
+        initializer_glorot = tf.keras.initializers.GlorotUniform(seed=random.randint(0, 1000000))
+
+        self.seq = Sequential([
+            Dense(units, kernel_initializer=initializer_glorot),
+            LayerNormalization()
+        ])
+
+    def call(self, x):
+        return self.seq(x)
+
+
+def add_residual_block(input, filters, kernel_size):
+    """
+    Add residual blocks to the model. If the last dimensions of the input data
+    and filter size does not match, project it such that last dimension matches.
+    """
+    out = ResidualMain(filters, kernel_size)(input)
+
+    res = input
+    # Using the Keras functional APIs, project the last dimension of the tensor to
+    # match the new filter size
+    if out.shape[-1] != input.shape[-1]:
+        res = Project(out.shape[-1])(res)
+
+    return add([res, out])
+
+
+class ResizeVideo(Layer):
```

Comment:
If this can't be refactored to not use einops we should put this layer outside in the examples directory and try to plug in into the model.